bugs/lp-1884071-gh-448.txt

```diff
@@ -0,0 +1,326 @@
+=== Begin SRU Template ===
+
+[IMPACT]
+Hetzner cloud only supports user-data as a string (presumabily utf-8).
+
+In order to allow users on Hetzner to provide binary data to cloud-init,
+we will attempt to base64decode the userdata.
+
+The change here adds a 'maybe_b64decode' function that will decode data
+if and only if is base64 encoded.
+
+[TEST CASE]
+#!/bin/bash
+
+set -uf -o pipefail
+
+# Redirect output to file
+exec > gh-448.log
+exec 2>&1
+
+export LC_ALL=C.UTF-8
+
+fail() {
+    echo "[VERIFICATION FAILED]"
+    exit 1
+}
+
+cleanup() {
+    [[ -n $name ]] || return
+    if hcloud server describe "$name" > /dev/null 2>&1; then
+        echo "Cleaning up instance $name"
+        hcloud server delete "$name"
+    fi
+}
+
+gh448() {
+    release=$1
+    name="gh448-$release"
+    keyname="paride"
+    sshopts=(-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR)
+
+    echo "=== BEGIN $release ==="
+
+    # Cleanup on exit
+    trap cleanup EXIT
+
+    # Just in case we have a stale instance
+    cleanup
+
+    # Start the instance
+    hcloud server create --name "$name" --user-data-from-file gh-448.yaml.b64 --ssh-key "$keyname" --image "$release" --type cx11 || fail
+
+    # Arbitrary, good enough for this validation
+    sleep 60
+
+    echo "--- Started instance $name with image release"
+
+    ip=$(hcloud server describe -o json "$name" | jq --raw-output '.public_net.ipv4.ip')
+
+    ssh "${sshopts[@]}" "root@$ip" -- cloud-init status --wait --long
+
+    # Hetzner runs a patched version of cloud-init that already implements b64 decoding,
+    # so /tmp/gh-448-verification should be already present
+    echo "--- Verify that Hetzner's custom cloud-init decodes b64"
+    ssh "${sshopts[@]}" "root@$ip" -- test -f /tmp/gh-448-verification || fail
+    echo "--- OK"
+
+    # Reinstall stock cloud-init. The file shouldn't be created now.
+    echo "--- Verify that pre-SRU cloud-init does not decode b64"
+    ssh "${sshopts[@]}" "root@$ip" -- apt-get -qq update
+    ssh "${sshopts[@]}" "root@$ip" -- DEBIAN_FRONTEND=noninteractive apt-get -qq install --reinstall cloud-init
+    ssh "${sshopts[@]}" "root@$ip" -- cloud-init clean --logs --reboot
+    sleep 60
+    ssh "${sshopts[@]}" "root@$ip" -- cloud-init status --wait --long
+    ssh "${sshopts[@]}" "root@$ip" -- test ! -f /tmp/gh-448-verification || fail
+    echo "--- OK"
+
+    # Add the -proposed repository, upgrade, reboot. The test file should now be present.
+    echo "--- Verify that -proposed cloud-init does decode b64"
+    ssh "${sshopts[@]}" "root@$ip"  add-apt-repository ppa:cloud-init-dev/proposed
```

Comment:
I was originally using the ppa as well to perform the tests, but as @blackboxsw pointed out in this review:
https://github.com/cloud-init/ubuntu-sru/pull/153#discussion_r485169456

I think we can move to using the proposed pocket for each release when testing